telegram/ext/_utils/webhookhandler.py

```diff
@@ -43,131 +40,126 @@
 
 
 class WebhookServer:
+    """Thin wrapper around ``tornado.httpserver.HTTPServer``."""
+
     __slots__ = (
-        'http_server',
+        '_http_server',
         'listen',
         'port',
-        'loop',
-        'logger',
+        '_logger',
         'is_running',
-        'server_lock',
-        'shutdown_lock',
+        '_server_lock',
+        '_shutdown_lock',
     )
 
     def __init__(
-        self, listen: str, port: int, webhook_app: 'WebhookAppClass', ssl_ctx: SSLContext
+        self, listen: str, port: int, webhook_app: 'WebhookAppClass', ssl_ctx: Optional[SSLContext]
     ):
-        self.http_server = HTTPServer(webhook_app, ssl_options=ssl_ctx)
+        self._http_server = HTTPServer(webhook_app, ssl_options=ssl_ctx)
         self.listen = listen
         self.port = port
-        self.loop: Optional[IOLoop] = None
-        self.logger = logging.getLogger(__name__)
+        self._logger = logging.getLogger(__name__)
         self.is_running = False
-        self.server_lock = Lock()
-        self.shutdown_lock = Lock()
+        self._server_lock = asyncio.Lock()
+        self._shutdown_lock = asyncio.Lock()
 
-    def serve_forever(self, ready: Event = None) -> None:
-        with self.server_lock:
-            IOLoop().make_current()
-            self.is_running = True
-            self.logger.debug('Webhook Server started.')
-            self.loop = IOLoop.current()
-            self.http_server.listen(self.port, address=self.listen)
+    async def serve_forever(self, ready: asyncio.Event = None) -> None:
+        async with self._server_lock:
+            self._http_server.listen(self.port, address=self.listen)
 
+            self.is_running = True
             if ready is not None:
                 ready.set()
 
-            self.loop.start()
-            self.logger.debug('Webhook Server stopped.')
-            self.is_running = False
+            self._logger.debug('Webhook Server started.')
 
-    def shutdown(self) -> None:
-        with self.shutdown_lock:
+    async def shutdown(self) -> None:
+        async with self._shutdown_lock:
             if not self.is_running:
-                self.logger.warning('Webhook Server already stopped.')
+                self._logger.debug('Webhook Server is already shut down. Returning')
                 return
-            self.loop.add_callback(self.loop.stop)  # type: ignore
+            self.is_running = False
+            self._http_server.stop()
+            await self._http_server.close_all_connections()
+            self._logger.debug('Webhook Server stopped')
 
     # pylint: disable=unused-argument
     def handle_error(self, request: object, client_address: str) -> None:
```

Comment:
apparently this method is not used?

Reply:
IIRC it's a methods that may be called directly by tornado. I wasn't completely sure if I had understood all of tornados setup correctly, so I kept all methods from before …

Reply:
but this class itself does not inherit from any tornado class? So how could it be called by tornado

Reply:
right. Got confused. Seems to be a remnant from the pre-tornado time - see #985, where `BaseHTTPServer` was still used. I'll double check & remove if possible :)